---
title: "NPRC Project - Data Preparation"
author: "Shaurita Hutchins"
date: "`r Sys.Date()`"
output: 
  mizzourahmd::mizzou_document:
    code_folding: "hide"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE, comment = "#>")
```


## What is Truffle?

### Input Data

Truffle requires biallelic data and WGS only data. If exome data is being used, parameters need to be tweaked. Initially, when the data included exome data, I attempted to tweak the parameters to no avail.

The output file for Truffle is very similar to the input file for the pedigree reconstruction pipeline except it has not been normalized. The normalization step does not seem to be necessary for Truffle. There was no need to filter or subset data by the minor allele frequency given that Truffle can filter data based on it.

```{bash, eval=FALSE}
DATA_DIR=/ddn/home3/vallender/projects/NPRC_Pedigree_Project/data

INFILE=${DATA_DIR}/mGAP/all_nprcs_rminfo.vcf.gz
OUTFILE=${DATA_DIR}/truffle_input/all_nprcs_rminfo_autosomal_wgs_biallelic_mt.vcf.gz
CHROMOSOMES=chr01,chr02,chr03,chr04,chr05,chr06,chr07,chr08,chr09,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19,chr20,MT
SAMPLES_FILE=/ddn/home3/vallender/projects/NPRC_Pedigree_Project/data/metadata/wgs_ids.txt

module use /ddn/home3/vallender/software/module_files
module load nprc_project

bcftools view $INFILE -r $CHROMOSOMES -S $SAMPLES_FILE -O z -o $OUTFILE -m2 -M2 -v snps,indels --threads 80

bcftools index $OUTFILE --threads 80

```

The script for this can be found at `/ddn/home3/vallender/projects/NPRC_Pedigree_Project/src/truffle_prep` on the MCSR.

There are two data files available for truffle: `all_nprcs_rminfo_autosomal_wgs_biallelic_mt.vcf.gz` and `all_nprcs_rminfo_autosomal_wgs_biallelic_un_mt.vcf.gz`. They are both located in `/ddn/home3/vallender/projects/NPRC_Pedigree_Project/data/truffle_input`. The two data files were generated in order to test whether `chrUn` (unknown snps) has an impact on the data.

### Running Truffle

```{bash, eval=FALSE}
#!/bin/bash
# ------------------------------------------------------------------
# [Vallender Lab] Truffle Relationship Estimation
#          Run truffle analysis on a vcf of all nprcs that's normalized.
#          VCF file with only wgs, biallelic, and MT variants.
#PBS -N truffle
#PBS -l select=15:ncpus=1:mem=4000mb
#PBS -l place=free
#PBS -j oe
#PBS -o logs
#PBS -e logs

VAR=all_nprcs_rminfo_autosomal_wgs_biallelic_mt
RESULTSDIR=/ddn/home3/vallender/projects/NPRC_Pedigree_Project/results/truffle/relationship_estimation/${VAR}
VCF=/ddn/home3/vallender/projects/NPRC_Pedigree_Project/data/truffle_input/${VAR}.vcf.gz

module use /ddn/home3/vallender/software/module_files
module load nprc_project

mkdir $RESULTSDIR

truffle --vcf $VCF --cpu 10 --mindist 5000 --nofiltering --out $RESULTSDIR/$VAR

chmod 0770 $RESULTSDIR/*
chmod 0770 $RESULTSDIR

```

### Results

We used the R package, [pophelper](https://github.com/royfrancis/pophelper), to help us visualize our data.
